app/src/main/java/jp/yhonda/MaximaOnAndroidActivity.java

```diff
@@ -860,4 +887,82 @@ private void sessionMenu(String cmd) {
 		editText.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_UP,
 				KeyEvent.KEYCODE_ENTER));
 	}
+
+	private void selectScriptFile() {
+		if(Build.VERSION.SDK_INT < 19) {
+			//Versions earlier than KitKat do not support the Storage Access Framework.
+			//Show file path input box instead:
+			android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(MaximaOnAndroidActivity.this);
+			LayoutInflater inflater = this.getLayoutInflater();
+			final View layout = inflater.inflate(R.layout.scriptpathalert, null);
+			builder.setView(layout);
+			final AutoCompleteTextView textView = (AutoCompleteTextView) layout.findViewById(R.id.scriptPathInput);
+			builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+				@Override
+				public void onClick(DialogInterface dialog, int which) {
+					dialog.dismiss();
+					if(!textView.getText().toString().matches("")) {
+						copyScriptFileToInputArea(Uri.fromFile(new File(textView.getText().toString())));
+					}
+				}
+			});
+			android.support.v7.app.AlertDialog dialog = builder.create();
+			dialog.show();
+
+		} else {
+			Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
+			intent.addCategory(intent.CATEGORY_OPENABLE);
+			intent.setType("*/*");
+			startActivityForResult(intent, READ_REQUEST_CODE);
+		}
+	}
+
+	private void copyScriptFileToInputArea(Uri fileUri) {
+        //Copy script file contents into a temporary file:
+        File temporaryDirectory = getApplicationContext().getCacheDir();
+        File temporaryFile = null;
+        try {
+            InputStream inputStream = getContentResolver().openInputStream(fileUri);
+            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
+            String scriptLine;
+            temporaryFile = File.createTempFile("userscript", ".mac", temporaryDirectory);
+            OutputStream outputStream = getContentResolver().openOutputStream(Uri.fromFile(temporaryFile));
+            BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(outputStream));
+            long writtenBytes = 0;
+            while((scriptLine = bufferedReader.readLine()) != null) {
+                writtenBytes += (scriptLine.length() + 1);
+                if (writtenBytes > 1E7) { //Limit the size of scripts to 10MB.
```

Comment:
My reason for limiting the size was that the file is copied into the temporary files. If the user accidentally selects a huge file (video, ...), the copying can take a long time. Also, I am not sure how well maxima handles huge files that are not text. </br> Since I am stuck with `InputStream`, I could not find a better way to check the file size before copying. It seems using `InputStream.available()` is generally discouraged for this purpose. If I find another way of handling files in general, this might be easier to do in the future.